src/MediaDeviceHandler.ts

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2017 Michael Telatynski <7t3chguy@gmail.com>
+Copyright 2021 Šimon Brandner <simon.bra.ag@gmail.com>
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import SettingsStore from "./settings/SettingsStore";
+import { SettingLevel } from "./settings/SettingLevel";
+import { setMatrixCallAudioInput, setMatrixCallVideoInput } from "matrix-js-sdk/src/matrix";
+import EventEmitter from 'events';
+
+interface IMediaDevices {
+    audioOutput: Array<MediaDeviceInfo>;
+    audioInput: Array<MediaDeviceInfo>;
+    videoInput: Array<MediaDeviceInfo>;
+}
+
+export enum MediaDeviceHandlerEvent {
+    AudioOutputChanged = "audio_output_changed",
+}
+
+export default class MediaDeviceHandler extends EventEmitter {
+    private static internalInstance;
+
+    public static get instance(): MediaDeviceHandler {
+        if (!MediaDeviceHandler.internalInstance) {
+            MediaDeviceHandler.internalInstance = new MediaDeviceHandler();
+        }
+        return MediaDeviceHandler.internalInstance;
+    }
+
+    public static async hasAnyLabeledDevices(): Promise<boolean> {
+        const devices = await navigator.mediaDevices.enumerateDevices();
+        return devices.some(d => Boolean(d.label));
+    }
+
+    public static async getDevices(): Promise<IMediaDevices> {
+        // Only needed for Electron atm, though should work in modern browsers
+        // once permission has been granted to the webapp
+
+        try {
+            const devices = await navigator.mediaDevices.enumerateDevices();
+
+            const audioOutput = [];
+            const audioInput = [];
+            const videoInput = [];
+
+            devices.forEach((device) => {
+                switch (device.kind) {
+                    case 'audiooutput': audioOutput.push(device); break;
+                    case 'audioinput': audioInput.push(device); break;
+                    case 'videoinput': videoInput.push(device); break;
+                }
+            });
+
+            return {
+                audioOutput: audioOutput,
+                audioInput: audioInput,
+                videoInput: videoInput,
+            };
```

Comment:
```suggestion
            return { audioOutput, audioInput, videoInput };
```

Just making sure you're aware of this short-hand, can be cleaner

Reply:
I am semi-aware of it, I seem to always forget about it :D